[Ho Zhan Rui,Glenn] iP

ghzr0 opened the pull request.

ghzr0 committed:
Made changes according to Level 0-2 requirements:
1. Added Greet
2. Added List
3. Added Exit

ghzr0 committed:
Updated to Level-3 increment :
1. Added the ability to mark & unmark tasks

Khulon commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,94 @@
+import java.util.Scanner;
+import java.util.ArrayList;
 public class Duke {
+
```
Well commented code

Khulon commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,94 @@
+
+    //Definition lists //
+    public static final String margin = "*----------------------------*";
+    public static String updateStr = "added: ";          // for adding new tasks in taskList
```
"str" is not so clear (readability)

Khulon replied:
"str" is not so clear (readability)

Khulon commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,94 @@
+        }
+
+    // How the program runs based on user's input commands //
+    public static void enter(String userCommand) {
```
indentation should be 8 spaces

Khulon commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,94 @@
+
+        String[] strNumSeparator = userCommand.split(" ");  // split string & number components
+
+        switch (strNumSeparator[0]) {
```
Good naming convention, follows camel case

Khulon commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,94 @@
+        }
+
+    // How the program runs based on user's input commands //
+    public static void enter(String userCommand) {
```
indentation should be 8 spaces

ghzr0 replied:
Hi @Khulon, could you elaborate why is there a need for 8 space indentation? As far as I know defining methods in a class only need a 4-space indentation.

Khulon commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,94 @@
+import java.util.ArrayList;
 public class Duke {
+
+    //Definition lists //
```
Well commented code

Khulon commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,94 @@
+            case "list":       // View taskList
+                System.out.println(margin);
+                System.out.println("Here are the tasks in your list:");
+                for (int i = 0; i < taskCount; i++) {
```
could use taskNumber instead of "i" to be clearer (for better readability)

Brennanzuz replied:
I think the coding standards mentioned how single letter iterators are allowed for loops. So I think this is okay :)

Khulon reviewed (commented):
Hi, good job on your code!

Brennanzuz commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,94 @@
+    public static final String margin = "*----------------------------*";
+    public static String updateStr = "added: ";          // for adding new tasks in taskList
+
+    public static Task[] taskList = new Task[100];  // Create array to store the tasks
```
The plural form should be used to represent the array, so `tasksArray` would follow the coding standard.

Brennanzuz commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,94 @@
 public class Duke {
+
+    //Definition lists //
+    public static final String margin = "*----------------------------*";
```
`final` variables are constants that should be capitalised, so `MARGIN` would be a better variable name.

Brennanzuz commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,94 @@
+                int taskToMark = Integer.parseInt(strNumSeparator[1]) - 1;      // identify which task to mark
+                System.out.println(margin);
+                System.out.println("Nice! I've marked this task as done:");
+                taskList[taskToMark].setDone(true);
```
Good and consistent use of setters even though it's tempting to directly access and modify.👍

Brennanzuz commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,94 @@
+
+        String[] strNumSeparator = userCommand.split(" ");  // split string & number components
+
+        switch (strNumSeparator[0]) {
+            case "bye":        //Exit the program
+                System.out.println(margin);
+                System.out.println("Bye. Hope to see you again soon!");
+                System.out.println(margin);
+                break;
+
+            case "list":       // View taskList
+                System.out.println(margin);
+                System.out.println("Here are the tasks in your list:");
+                for (int i = 0; i < taskCount; i++) {
+                    System.out.println(Integer.toString(i + 1) + ". " + "["+ taskList[i].getStatusIcon() + "]" + taskList[i] );
+                }
+                System.out.println(margin);
+                break;
+
+            case "mark":       // Mark task as done
+                int taskToMark = Integer.parseInt(strNumSeparator[1]) - 1;      // identify which task to mark
+                System.out.println(margin);
+                System.out.println("Nice! I've marked this task as done:");
+                taskList[taskToMark].setDone(true);
+                System.out.println("  [" +  taskList[taskToMark].getStatusIcon() + "] " + taskList[taskToMark]);
+                System.out.println(margin);
+                break;
+
+
+            case "unmark":      // unmark task
+                int taskToUnmark = Integer.parseInt(strNumSeparator[1]) - 1;      // identify which task to un
+                System.out.println(margin);
+                System.out.println("OK, I've marked this task as not done yet:");
+                taskList[taskToUnmark].setDone(false);
+                System.out.println("  [" +  taskList[taskToUnmark].getStatusIcon() + "] " + taskList[taskToUnmark]);
+                System.out.println(margin);
+                break;
+
+            default:            // adding new tasks
```
The switch statement's cases should all be on the same level of indentation (weird I know, but it said so here [https://se-education.org/guides/conventions/java/basic.html](https://se-education.org/guides/conventions/java/basic.html) :/)

Brennanzuz commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    public String getStatusIcon() {
+        return (isDone ? "X" : " "); // mark done task with X
```
Not entirely sure if this was allowed, but I'd rather not use it and stick with if-else statements for consistency sake.

Brennanzuz reviewed (commented):
Has a few minor coding standard violations but otherwise it is mostly acceptable at this stage.

ghzr0 committed:
Add Increments: Level-4, A-CodeQuality
1. Implemented ToDos, Events, Deadlines
2. Edited to fit coding quality and standards based on PRs

ghzr0 committed:
1. Added increment Level 5 to handle errors inputted by user
2. Created DukeException class to store and output error scenarios and messages
3. Repackaged and reorganised previous code with methods to facilitate readability

ghzr0 committed:
Merge branch 'branch-Level-5'

ghzr0 committed:
Organised classes into suitable packages ( Duke, duke.task , duke.exception)

ghzr0 committed:
Added Level-6 increment:
1. delete function supported

ghzr0 committed:
Added increment Level-7
1. Save file
Additional changes:
1. Created Storage and UI classes for better packaging

ghzr0 committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Task.java
#	src/main/java/duke/UI.java
#	src/main/java/duke/storage/Storage.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/task/Todo.java

ghzr0 committed:
Merged Level-6 and Level-7 branches to master branch.
Additonal commits:
1. Debug some lines
2.Implemented .jar file to be able to open in a new empty folder

ghzr0 committed:
Refactor code to integrate more OOP, added new packages and classes to group relevant methods and functions together

ghzr0 committed:
no message

ghzr0 committed:
Reorganised code and integrated OOP to hold the different classes and functions

ghzr0 committed:
Commited increment Level-9
1. Added find function
2. Updated Parser class to include 'find' case
3. Updated commands package to include FindCommand() class

ghzr0 committed:
Reorganised code formats and included more functions to cover edge test cases

ghzr0 committed:
Updated filePath to store the textfile

ghzr0 committed:
Fixed some errors and debugging

ghzr0 committed:
Update README.md

ghzr0 committed:
Updated UserGuide for Duke Program

ghzr0 committed:
Merge branch 'master' of https://github.com/ghzr0/ip

* 'master' of https://github.com/ghzr0/ip:
  Updated UserGuide for Duke Program
  Update README.md

chinhan99 commented on src/main/java/duke/commands/EventCommand.java:
```diff
@@ -0,0 +1,46 @@
+import duke.tasks.Event;
+import duke.outputs.Messages;
+import duke.file.TaskList;
+import duke.ui.*;
```
avoid using star imports !

chinhan99 commented on src/main/java/duke/commands/EventCommand.java:
```diff
@@ -0,0 +1,46 @@
+    private Event event;
+
+    /**
+     * creates a new event task
```
Do check the capitalisation and use of periods in your comments are in line with the coding standards. 

https://se-education.org/guides/conventions/java/basic.html#comments

chinhan99 commented on src/main/java/duke/commands/MarkCommand.java:
```diff
@@ -0,0 +1,46 @@
+
+import duke.tasks.Task;
+import duke.file.TaskList;
+import duke.ui.*;
```
star import. Please update this too.

chinhan99 commented on src/main/java/duke/outputs/Messages.java:
```diff
@@ -0,0 +1,124 @@
+    public static final String EXIT_MESSAGE = "Bye. Hope to see you again soon!";
+
+    //This block is for the possible errors that may occur during user inputs
+    public static final String ERROR_MESSAGE_EMPTY_LIST = "The current task list is empty.";
+    public static final String ERROR_MESSAGE_WRONG_TASK_INDEX = "Please specify a valid task number!";
+    public static final String ERROR_MESSAGE_UNKNOWN_COMMAND = "☹ OOPS!!! I do not understand this command. :-(";
+    public static final String ERROR_MESSAGE_EMPTY_TODO = "☹ OOPS!!! The description of todo cannot be empty.";
+    public static final String ERROR_MESSAGE_EMPTY_DEADLINE = "☹ OOPS!!! The description of deadline cannot be empty.";
+    public static final String ERROR_MESSAGE_EMPTY_EVENT = "☹ OOPS!!! The description of event cannot be empty.";
+    public static final String ERROR_MESSAGE_EMPTY_DELETE = "Please specify a task number to delete!";
+    public static final String ERROR_MESSAGE_EMPTY_MARK = "Please specify a task number to mark!";
+    public static final String ERROR_MESSAGE_EMPTY_UNMARK = "Please specify a task number to unmark!";
+    public static final String ERROR_MESSAGE_NON_NUMERICAL_INDEX_INPUT = "Please key in a valid numerical input as the task index!";
+    public static final String ERROR_MESSAGE_KEYWORD_NOT_FOUND = "Oops! I could not find any task matching that keyword.";
+    public static final String ERROR_MESSAGE_KEYWORD_UNSPECIFIED = "Please enter a keyword to for me to search.";
+    public static final String ERROR_MESSAGE_FILE_IOEXCEPTION_ERROR = "☹ OOPS!!! The file is corrupted.";
+    public static final String ERROR_MESSAGE_FILE_LOAD_ERROR = "There is an error during loading of the file.";
```
Good use of constants to contain commonly used messages.

chinhan99 commented on src/main/java/duke/outputs/Messages.java:
```diff
@@ -0,0 +1,124 @@
+    public static final String ERROR_MESSAGE_FILE_LOAD_ERROR = "There is an error during loading of the file.";
+
+    // Some common error messages packaged into functions for easier calling
+    public static void emptyTodoErrorMessage() {
```
Do note that method names should be in verb form.

https://se-education.org/guides/conventions/java/basic.html#comments

chinhan99 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,90 @@
+
+        case "deadline":
+            command = new DeadlineCommand();
+            command.execute(input, tasks, ui);
+            break;
+
+        case "event":
+            command = new EventCommand();
+            command.execute(input, tasks, ui);
+            break;
+
+        case "list":
+            command = new ListCommand();
+            command.execute(input, tasks, ui);
+            break;
+
+        case "bye":
+            UI.endProgram();
+            isByeCommand = true;
+            break;
+
+        case "mark":
+            command = new MarkCommand();
+            command.execute(input, tasks, ui);
+            break;
+
+        case "unmark":
+            command = new UnmarkCommand();
+            command.execute(input, tasks, ui);
+            break;
+
+        case "delete":
+            command = new DeleteCommand();
+            command.execute(input, tasks, ui);
+            break;
+
+        case "find":
+            command = new FindCommand();
+            command.execute(input, tasks, ui);
+            break;
```
Good abstraction of submethods

chinhan99 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,125 @@
+
+    /**
+     * Creates a new text file if it does not exist
+     *
+     * @throws IOException if error during file creation
+     */
+    public void createTextFile() throws IOException {
+        File file = new File(filePath);
+        if (!file.exists()) {
+            file.createNewFile();
+            file.mkdirs();
+        }
+    }
+
+    /**
+     * Updates tasklist with data from existing save file if it exists
+     *
+     * @param tasks the tasklist to be updated
+     */
+    public void loadTextFile(TaskList tasks) {
+        try {
+            BufferedReader reader = new BufferedReader(new FileReader(filePath));
+            String textFileLine;
+            while ((textFileLine = reader.readLine()) != null) {
+                String[] textLinesArray = textFileLine.split("\\|", -1);
+                if (textFileLine.startsWith("T")) {
+                    Todo todo = new Todo(textLinesArray[2]);
+                    if (textLinesArray[1].equals("1")) {
+                        todo.markAsDone();
+                    }
+                    tasks.addNewTask(todo);
+                } else if (textFileLine.startsWith("D")) {
+                    Deadline deadline = new Deadline(textLinesArray[2], textLinesArray[3]);
+                    if (textLinesArray[1].equals("1")) {
+                        deadline.markAsDone();
+                    }
+                    tasks.addNewTask(deadline);
+                } else {
+                    Event event = new Event(textLinesArray[2], textLinesArray[3], textLinesArray[4]);
+                    if (textLinesArray[1].equals("1")) {
```
Do consider replacing magic literals with constants

chinhan99 reviewed (commented):
Good OOP structure used within this project. Do note the coding standards required as there were methods that were incorrectly named and magic literals used .

ghzr0 committed:
Update README.md

ghzr0 committed:
Update README.md

ghzr0 committed:
Update README.md

ghzr0 committed:
Update README.md

ghzr0 committed:
Removed unused import statements and wildcard imports

ghzr0 committed:
Merge branch 'master' of https://github.com/ghzr0/ip
